classes/Product.php

```diff
@@ -2626,39 +2626,26 @@ public function getAttributeCombinationsById($id_product_attribute, $id_lang, $g
     }
 
     /**
+     * Returns information about product images that are paired to specific combination
+     *
      * @param int $id_lang Language identifier
      *
      * @return array|false
      */
     public function getCombinationImages($id_lang)
     {
+        // If combination feature is disabled, no need to do any queries
         if (!Combination::isFeatureActive()) {
             return false;
         }
 
-        $product_attributes = Db::getInstance()->executeS(
-            'SELECT `id_product_attribute`
-            FROM `' . _DB_PREFIX_ . 'product_attribute`
-            WHERE `id_product` = ' . (int) $this->id
-        );
-
-        if (!$product_attributes) {
-            return false;
-        }
-
-        $ids = [];
-
-        foreach ($product_attributes as $product_attribute) {
-            $ids[] = (int) $product_attribute['id_product_attribute'];
-        }
-
         $result = Db::getInstance()->executeS(
             '
             SELECT pai.`id_image`, pai.`id_product_attribute`, il.`legend`
             FROM `' . _DB_PREFIX_ . 'product_attribute_image` pai
             LEFT JOIN `' . _DB_PREFIX_ . 'image_lang` il ON (il.`id_image` = pai.`id_image`)
             LEFT JOIN `' . _DB_PREFIX_ . 'image` i ON (i.`id_image` = pai.`id_image`)
```

Comment:
Just my 2cents:
'.Shop::addSqlAssociation('product_attribute', 'pai').'

In front there is no need to bring back images that do not concern the current shop.

Reply:
From experience, the transfer time is virtually nothing even for huge lists of data. See what we did with faceted search to speed it up twice.

Adding that join would bring no speed gain I think. And on most multistore shops, the images will be the same on all stores, so it would be just an extra join.

Reply:
The join is mainly used to avoid bringing back images corresponding to attributes that do not exist on a shop.
I have many examples of multishop (B2B/B2C) where the attributes are not the same.
Why bring back images that will not correspond to any attribute?
This array is loaded into memory and then processed in php afterwards.

An INNER costs nothing in SQL, but that's just my opinion.

Reply:
> Adding that join would bring no speed gain I think. **And on most multistore shops, the images will be the same on all stores**, so it would be just an extra join.

**You don't know that.**

Reply:
I don't, but I got rid of one query and potentialy sped up the second one.

If someone can improve it, another PR can follow.

Reply:
And
LEFT JOIN `' . _DB_PREFIX_ . 'image` i ON (i.`id_image` = pai.`id_image`)
should be
INNER JOIN `' . _DB_PREFIX_ . 'image` i ON (i.`id_image` = pai.`id_image`)

Reply:
Yeah, but it should work the same. It would exclude the image only in case of database inconsistency. I can add this, I thought about it before. :-)

Also, maybe the legend can go, if its not needed on another places.

Reply:
But, even in current form, the PR has zero performance drawbacks, in all configuration. This discussion about shop join is only a potential improvement, for small fraction of people. And potential drawback for the majority. I will think about it.

Reply:
@Eolia Inner join added. :-)